File: UnitConversionLibrary/CPP/UnitConversion/Source/ConversionBase.cpp
/////////////////////////////////////////////////////////////////////////////////
//2345678901234567890123456789012345678901234567890123456789012345678901234567890
/////////////////////////////////////////////////////////////////////////////////
// PROJECT: Unit Conversion
//
// Copyright Copyright 2024 MAP
//
// Unpublished - Rights reserved under the Copyright Laws of the United States
//
/////////////////////////////////////////////////////////////////////////////////
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful, 
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
//
/////////////////////////////////////////////////////////////////////////////////
//
// File ConversionBase.cpp
//
// Base class for conversions. This class provides methods for accessing
// the units, systems, and type groups that constitute a system of units 
// for a particular unit type. 
//
// Version 1.0
// 
/////////////////////////////////////////////////////////////////////////////////
#include "ConversionBase.hpp"

/**
 * @brief Default constructor.
 */
ConversionBase::ConversionBase(void) :  _map(), 
                                        _name("Invalid"), 
                                        _valid(false)
{
}

/**
 * @brief Constructor.
 *
 * @param[in] name - the name of the conversion base.
 */
ConversionBase::ConversionBase(const std::string& name) :  _map(), 
                                                           _name(name), 
                                                           _valid(true)
{
}

/**
 * @brief Constructor.
 *
 * @param[in] name - the name of the conversion base.
 * @param[in] map  - the units in the conversion base.
 */
ConversionBase::ConversionBase(const std::string& name, 
                               const std::map<std::string, 
                               BaseSystem>& map) : _map(map), 
                                                   _name(name), 
                                                   _valid(true)
{
}

/**
 * @brief Copy constructor.
 *
 * @param[in] other - the ConversionBase object to copy.
 */
ConversionBase::ConversionBase(const ConversionBase& other) :  _map(other._map),
                                                               _name(other._name), 
                                                               _valid(other._valid)
{
}

/**
 * @brief Assignment Operator.
 *
 * @param[in] other - the ConversionBase object to copy.
 */
ConversionBase& ConversionBase::operator=(const ConversionBase& other)
{
    _map   = other._map;
    _name  = other._name;
    _valid = other._valid;
    return *this;
}

/**
  * @brief Construct a actual type name. Type name is dependent on
  *        on the system: either CanonicalSystem or SingleSystem.
  * 
  * @param[in] type   - the type.
  * @param[in] system - the system.
  * 
  * @return - system specific type.
  */
std::string ConversionBase::actualType(const std::string& type, 
                                       const std::string& system) const
{
    return type;
}

/*
 * @brief return a list of type or system names in the BaseSystem
 * 
 * @param[in] bs - BaseSystem whose names are required.
 * 
 * @ return - a list of names.
 */
std::list<std::string> ConversionBase::baseSystemNames(const BaseSystem& bs) const
{
    return std::list<std::string>();
}

/**
 * @brief The check method checks if the software version matches
 *        with the unit versions of its units.
 * 
 * @return - true if software and unit versions match, false otherwise.
 */
bool ConversionBase::check(void) const
{
    for (std::map<std::string, BaseSystem>::const_iterator it = _map.begin(); 
                                                           it != _map.end(); ++it)
    {
        BaseSystem bs = it->second;
        if (bs.check())
        {
            continue;
        }
        else
        {
            return false;
        }
    }
    return true;
}

/**
 * @brief Check is system is part of unit name. 
 *
 * @param[in] sysName  - the system name.
 * @param[in] unitName - the unit name.
 *
 * @return - the full unit name.
 */ 
bool ConversionBase::checkName(const std::string& sysName, 
                               const std::string& unitName) const
{
    if (unitName.find(sysName) != 0)
    {
        return true;
    }
    else
    {
        return false;
    }
}

/*
 * @brief return a list of type or system names.
 * 
 * @ return - a list of names.
 */
std::list<std::string> ConversionBase::coreNames(void) const
{
    std::list<std::string> keys;
    for (std::map<std::string, BaseSystem>::const_iterator it = _map.begin(); 
                                                           it != _map.end(); ++it) 
    {
        singleCoreNames(it->first, keys);
    }
    return keys;
}

/**
 * @brief Construct a full unit name from system and unit name. A full  
 *        unit name is of the form systemName[unitName] e.g. SI[newton].
 *        This method is overridden in the child class.
 *
 * @param[in] sysName  - the system name.
 * @param[in] unitName - the unit name.
 *
 * @return - the full unit name.
 */ 
std::string ConversionBase::fullName(const std::string& sysName, 
                                     const std::string& unitName) const
{
    std::string fullName = sysName + "[" + unitName + "]";
    return fullName;
}

/**
 * @brief Construct a full type name. Type name is dependent on
 *        on the system: either CanonicalSystem or SingleSystem.
 * 
 * @param[in] type   - the type.
 * @param[in] system - the system.
 * 
 * @return - system specific type.
 */
std::string ConversionBase::fullType(const std::string& type, 
                                     const std::string& system) const
{
    return type;
}

/**
 * @brief Construct the full unit name from the system and unit names.
 * 
 * @param[in] systemName - the system name.
 * @param[in] unitName   - the unit name.
 *
 * @return - the full unit name (unit name plus the system name).
 */
std::string ConversionBase::fullUnitName(const std::string& systemName, 
                                         const std::string& unitName) const
{
    return unitName;
}

/**
 * @brief Construct the local unit name from the system and unit names.
 * 
 * @param[in] systemName - the system name.
 * @param[in] unitName   - the unit name.
 *
 * @return - the full unit name (unit name plus the system name).
 */
std::string ConversionBase::localName(const std::string& systemName, 
                                      const std::string& unitName) const
{
    return unitName;
}

/**
 * @brief Construct a local type name. 
 *
 * param[in]  - type - the unit type.
 * 
 * @return - type for the unit based on class.
 */
std::string ConversionBase::localType(const std::string& type) const
{
    return type;
}

/*
 * @breif Construct list of map names.
 * 
 * @retrun a list of map keys.
 */
std::list<std::string> ConversionBase::mapNames(void) const
{
    std::list<std::string> keys;
    for (std::map<std::string, BaseSystem>::const_iterator it = _map.begin(); 
                                                           it != _map.end(); ++it) 
    {
        keys.push_back(it->first);
    }
    return keys;
}

/*
 * @brief Get the name of the conversion.
 *
 * @return the conversion name.
 */
std::string ConversionBase::name(void) const
{
    return _name;
}

/**
 * @brief Get the unit name from the full unit name.
 *
 * @param[in] unitName - the full unit name.
 *
 * @return - the unit name stripped of the full name.
 */
std::string ConversionBase::rawUnitName(const std::string& unitName) const
{
    size_t first = unitName.find('[');
    size_t last = unitName.find(']');
    first = (first == std::string::npos ? 0 : first+1);
    last = (last == std::string::npos ? unitName.length() : last);
    std::string uName = unitName.substr (first,last-first);
    return uName;
}

/*
 * @brief return a list of type or system names.
 * 
 * @param[in] name - unit type or system name.
 * @param[i] list  - list to place results into.
 */
void ConversionBase::singleCoreNames(const std::string& name, 
                                     std::list<std::string>& list) const
{
    if (_map.count(name) == 0)
    {
        return;
    }
    else
    {
        BaseSystem us = _map.at(name);
        std::list<std::string> names = baseSystemNames(us);
        for (std::list<std::string>::const_iterator i = names.begin(); 
                                                    i != names.end(); ++i)
        {
            bool found = (std::find(list.begin(), list.end(), *i) != list.end());
            if(found)
            {
                continue;
            }
            else
            {
                list.push_back(*i);
            }
        }
    }
}

/**
 * @brief Get a BaseSystem in the conversion.
 *
 * @ param[in] name - the name of the requested BaseSystem.
 *
 * @return the specified BaseSystem or an invalid BaseSystem if the
 *         requested BaseSystem was not found.
 */
BaseSystem ConversionBase::system(const std::string& name) const
{
    if (_map.count(name) == 0)
    {
        return BaseSystem();
    }
    else
    {
        return _map.at(name);
    }
}

/**
 * @brief Get a list of system names in the conversion. This method
 *        is overridden in the child class.
 *
 * @param[in] type - the unit type.
 *
 * @return - list of system names in the conversion.
 */ 
std::list<std::string> ConversionBase::systemNames(const std::string& type) const
{
    return std::list<std::string>();
}

/**
 * @brief Get a list of all system names in the conversion.
 *
 * @return - list of system names in the conversion.
 */ 
std::list<std::string> ConversionBase::systemNames(void) const
{
    return std::list<std::string>();
}

/**
 * @brief Get a type group from the conversion. This method is 
 *        overridden in the child class.
 *
 * @param[in] type   - the unit type.
 * @param[in] tgroup - the type group.
 *
 * @return - the requested type group or an invalid type 
 *           group if the unit system was not found.
 */ 
TypeGroup ConversionBase::typeGroup(const std::string& type, 
                                    const std::string& tgroup) const
{
    if (_map.count(actualType(type,tgroup)) == 0)
    {
        return TypeGroup();
    }
    else
    {
        TypeGroup tg = _map.at(actualType(type,tgroup)).typeGroup(localType(type));
        typeGroupFilter(tgroup, tg);
        return tg;
    }
}

/**
 * @brief filter TypeGroup
 * 
 * @param[in] name - type group name to keep.
 * @param[in] tg   - type group to filter.
 */
void ConversionBase::typeGroupFilter(const std::string& name, 
                                     TypeGroup& tg) const
{
    return;
}



/**
 * @brief Get a list of type names in the conversion. This
 *        method is overridden in the child class.
 *
 * @return - list of type name in the conversion.
 */ 
std::list<std::string> ConversionBase::typeNames(void) const
{
    return std::list<std::string>();
}

/**
 * @brief Get a unit from the conversion. This method is 
 *        overridden in the child class.
 *
 * @param[in] type   - the unit type.
 * @param[in] system - the unit system.
 * @param[in] name   - the unit name.
 *
 * @return - the requested unit or an invalid unit if the
 *           unit was not found.
 */ 
UBASE ConversionBase::unit(const std::string& type, 
                           const std::string& system, 
                           const std::string& name) const
{
    std::string typeName = actualType(type, system);
    if (_map.count(typeName) == 0)
    {
        return UBASE();
    }
    else
    {
        std::string unitName = fullName(system, name);
        std::string fullTypeName = fullType(type, system);
        return _map.at(typeName).unit(fullTypeName, unitName);
    }
}

/**
 * @brief Get a list of unit names in the conversion. This method
 *        is overridden in the child class.     
 *
 * @param[in] type   - the unit type.
 * @param[in] system - the unit system.
 *
 * @return - a list of unit names in the conversion.
 */ 
std::list<std::string> ConversionBase::unitNames(const std::string& type, 
                                                 const std::string& system) const
{
    std::list<std::string> keys;
    if (_map.count(actualType(type,system)) == 0)
    {
        return keys;
    }
    else
    {
        BaseSystem us = _map.at(actualType(type,system));
        std::list<std::string> names = us.unitNames(localType(type));
        for (std::list<std::string>::const_iterator i = names.begin(); 
            i != names.end(); ++i)
        {
            if(checkName(system, *i))
            {
                continue;
            }
            else
            {
                keys.push_back(localName(type, *i));
            }
        }
        return keys;
    }
}

/**
 * @brief Get a list of all unit names in the conversion. This method
 * is overridden in the child class.     
 *
 * @return - a list of all unit names in the conversion.
 */ 
std::list<std::string> ConversionBase::unitNames(void) const
{
    std::list<std::string> keys;
    for (std::map<std::string, BaseSystem>::const_iterator it = _map.begin(); 
        it != _map.end(); ++it) 
    {
        BaseSystem us = it->second;
        std::list<std::string> types = us.typeNames();
        for (std::list<std::string>::iterator itl = types.begin(); 
                                              itl != types.end(); ++itl)
        {
            std::list<std::string> names = us.unitNames(*itl);
            for (std::list<std::string>::const_iterator i = names.begin(); 
                i != names.end(); ++i)
            {
                bool found = (std::find(keys.begin(), keys.end(), *i) != keys.end());
                if (found)
                {
                    continue;
                }
                else
                {
                    keys.push_back(localName(*itl,*i));
                }
            }
        }
    }
    return keys;
}

/**
 * @brief Get the validity of the conversion.
 *
 * @return true if the conversion is valid, false otherwise.
 */
bool ConversionBase::valid(void) const
{
    return _valid;
}
// EOF

File: UnitConversionLibrary/Python/UnitConversion/ConversionBase.py
#################################################################################
#12345678901234567890123456789012345678901234567890123456789012345678901234567890
#################################################################################
# PROJECT: Unit Conversion
#
# Copyright Copyright 2024 MAP
#
# Unpublished - Rights reserved under the Copyright Laws of the United States
#
#################################################################################
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful, 
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
#
#################################################################################
#
# File ConversionBase.py
#
# Base class for conversions. This class provides methods for accessing
# the units, systems, and type groups that constitute a system of units 
# for a particular unit type. 
#
# Version 1.0
# 
#################################################################################
from . import BaseSystem
from . import TypeGroup
from . import UBASE
import copy

class ConversionBase:
    """
    Base class for conversions.
    
    Attributes
    ----------
        
    _name  - name of the Conversion object; typically the unit type.
    
    _map   - a Conversion consists of one or more BaseSystem with each 
             BaseSystem containing one or more type groups each containing one 
             or more units.
    
    _valid - a flag indicating if the Conversion is valid or not i.e. it 
             contains one or more valid BaseSystems.
                  
    Methods
    -------
    __init__        -  constructor.
    
    actualType      - construct the actual type name.
    
    baseSystemNames - get a list of type or system names in the BaseSystem.
    
    check           - check that software and uit versions match.

    checkName       - check is system is part of unit name.
    
    coreNames       - get a list of type or system names.
    
    fullName        - construct a full unit name from system and unit name.
    
    fullType        - construct a full type name. 
    
    fullUnitName    - construct a full unit name from system and unit name. 
    
    localName       - construct the local unit name from the system and unit names.
    
    localType       - construct a local type name. 
    
    mapNames        - construct list of map names.   

    name            - gets the name of the conversion.
    
    rawUnitName     - get the unit name from the full unit name.
    
    singleCoreNames - get a list of type or system names.
    
    system          - get a BaseSystem in the conversion.
    
    systemNames     - gets a list of all system names in the conversion.
    
    typeGroup       - gets a type groups from the system.
    
    typeGroupFilter - filter TypeGroup
    
    typeNames       - gets a list of type names in the conversion.
    
    unit            - gets a unit from the conversion.
    
    unitNames       - gets a list of all unit names in the conversion.
    
    valid           - gets the validity of the conversion.
    """
    
    def __init__(self, *inp):  
        """
        Constructor. If one or two arguments are supplied a valid Conversion 
        is constructed, otherwise a invalid Conversion is created.
        
        Parameters
        ----------
        For one argument:
        inp[0] name  - (input, string) name of the Conversion.

        OR:
        inp[0] other - (input, Conversion) other Conversion to copy.

        
        For two arguments:
        inp[0] name - (input, string) name of the Conversion.
        inp[1] map  - (input, dict) units in the Conversion.        
        """        
        if len(inp) == 1:
            if isinstance(inp[0],str):
                self._map = dict()
                self._name = inp[0]
                self._valid = True
                self._us = BaseSystem.BaseSystem()
            elif isinstance(inp[0],ConversionBase):
                self._map = copy.deepcopy(inp[0]._map)
                self._name = inp[0]._name
                self._valid = True
                self._us = BaseSystem.BaseSystem()        
            else:
                self._map = dict()
                self._name = "Invalid"
                self._valid = False
                self._us = BaseSystem.BaseSystem()            
        elif len(inp) == 2:
            self._map = copy.deepcopy(inp[1])
            self._name = inp[0]
            self._valid = True
            self._us = BaseSystem.BaseSystem()        
        else:
            self._map = dict()
            self._name = "Invalid"
            self._valid = False
            self._us = BaseSystem.BaseSystem()

    def actualType(self, typeName, systemName):
        """
        Construct the actual type name. Type name is dependent on  
        on the system: either CanonicalSystem or SingleSystem.
        This method is overridden in the child class.

        Parameters
        ----------
        typeName    - (input, string) the type.
        systemName  - (input, string) the system.

        Return
        ------
        The system specific type.
        Return type is string.
        """
        return typeName

    def baseSystemNames(self, bs):
        """
        Get a list of type or system names in the BaseSystem.
        
        Parameters
        ----------
        bs   - (input, BaseSystem) BaseSystem whose names are required.

        Return
        ------
        A list of names in the BaseSystm. Return type is list of strings.
        """
        return list()
            
    def check(self):
        """
        The check method checks if the software version matches
        with the unit versions of its units.
        
        Return
        ------
        True if software and unit versions match, false otherwise.
        """
        for item in self._map.items():
            if item[1].check() == True:
                continue
            else:
                return False
        return True
        
    def checkName(self, sysName, unitName):
        """
        Check is system is part of unit name. 
        
        Parameters
        ----------
        sysName  - (input, string) unit system name.
        unitName - (input, string) unit name.
        
        Return
        ------
        A boolean: true is system name is part of unit name.
        """
        if unitName.find(sysName) == 0:
            return False
        else:
            return True
            
    def coreNames(self):
        """
        Get a list of type or system names.
        
        Returns
        -------
        A list of type or system names. Return type is list of strings.
        """
        keys = []
        for item in self._map.items():
            self.singleCoreNames(item[0], keys)
        return keys
        

    def fullName(self,sysName, unitName):
        """
        Construct a full unit name from system and unit name. A full  
        unit name is of the form systeName[unitName] e.g. SI[newton].
        This method is overridden in the child class.

        Parameters
        ----------
        sysName  - (input, string) the system name.
        unitName - (input, string) the unit name.

        Return
        ------
        The full unit name as a string.
        Return type is string.
        """
        name = sysName + "[" + unitName + "]"
        return name

    def fullType(self,typeName, systemName):
        """
        Construct a full type name. Type name is dependent on  
        on the system: either CanonicalSystem or SingleSystem.
        This method is overridden in the child class.

        Parameters
        ----------
        typeName    - (input, string) the type.
        systemName  - (input, string) the system.

        Return
        ------
        The system specific type.
        Return type is string.
        """
        return typeName
        
    def fullUnitName(self, systemName, unitName):
        """
        Construct the full unit name from the system and unit names.

        Parameters
        ----------
        systemName  - (input, string) the system name.
        unitName    - (input, string) the unit name.

        Return
        ------
        The full unit name (unit name plus the system name).
        Return type is string.
        """
        return unitName
        
    def localName(self, systemName, unitName):
        """
        Construct the local unit name from the system and unit names.
        
        Parameters
        ---------
        systemName - (input, string) the system name.
        unitName   - (input, string) the unit name.
        
        Returns
        -------
        The full unit name (unit name plus the system name). 
        Return type is string.
        """
        return unitName
        
    def localType(self, tpe):
        """
        Construct a local type name. 
        
        Parameters
        ----------
        tpe - (input, string) the unit type.
        
        Returns
        -------
        Type for the unit based on class. 
        Return type is string.
        """
        return tpe
        
    def mapNames(self):
        """
        Construct list of map names.
        
        Returns
        -------
        A list of map keys.
        Return type is a list of strings.
        """
        keys = []
        for item in self._map.items():
            keys.append(item[0])
        return keys

    def name(self):
        """
        Get conversion name.
        
        Returns
        -------
        The name of the conversion.
        """
        return self._name

    def rawUnitName(self, unitName):
        """
        Get the unit name from the full unit name.

        Parameters
        ----------
        unitName    - (input, string) the full unit name.

        Return
        ------
        The unit name stripped of the system name as a string.
        Return type is string.
        """        
        try:
            first = unitName.index('[') + 1
            last = unitName.index(']')
        except:
            first =  0 
            last =  len(unitName) 
        size = last - first
        uName = unitName[first : first + size]
        return uName;

    def singleCoreNames(self,name, lst):
        """
        Get a list of type or system names.
        
        Parameters
        ----------
        name - (input, string)        unit type or system name.
        list - (output, list<string>) list to place results into.
        """
        if name in self._map.keys():
            us = self._map[name]
            names = self.baseSystemNames(us)
            for i in names: 
                if i in lst:
                    continue
                else:
                    lst.append(i)
        else:
            return

    def system(self, name):
        """
        Get a BaseSystem in the conversion.
        
        Parameters
        ----------
        name - name of the requested BaseSystem.

        Return
        ------
        The specified BaseSystem or an invalid BaseSystem if the
        requested BaseSystem was not found.
        Return type is BaseSystem.
        """
        if name in self._map.keys():
            return self._map[name]
        else:
            return self._us
        
    def systemNames(self, tpe = None):
        """
        Get a list of system names in the conversion. This method
        is overridden in the child class. 

        Parameters
        ----------
        tpe   - (input, string) the unit type.

        Return
        ------
        A list of system names in the conversion.
        Return type is a list of strings.
        """
        return list()
 
    def typeGroup(self, tpe, tgroup):
        """               
        Get a type group from the conversion. This method is overridden
        in the child class.

        Parameters
        ----------
        tpe    - (input, string) the unit type.
        tgroup - (input, string) the type group name.

        Return
        ------
        The requested type group or an invalid type 
        group if the type group was not found.
        Return type is TypeGroup.
        """
        actType = self.actualType(tpe,tgroup)
        if actType in self._map.keys():
            bs = self._map[actType]
            tg = TypeGroup.TypeGroup(bs.typeGroup(self.localType(tpe)))
            self.typeGroupFilter(tgroup, tg)
            return tg
        else:
            return TypeGroup.TypeGroup()
 
    def typeGroupFilter(self, name, tg):
        """
        Filter TypeGroup.
        
        Parameters
        ----------
        name - (input, string) type group name to keep.
        tg   - (output, TypeGroup) -  type group to filter.
        """
        return
    
    def typeNames(self):
        """
        Get a list of type names in the conversion. This
        method is overridden in the child class.

        Return
        ------
        A list of type name in the conversion.
        Return type is a list of strings.
        """
        return list()

    def unit(self, tpe, system, name):
        """
        Get a unit from the system. This method is overridden
        in the child class.

        Parameters
        ----------
        tpe    - (input, string) the unit type.
        system - (input, string) the unit system.
        name   - (input, string) the unit name.

        Return
        ------
        The requested unit or an invalid unit if the
        unit was not found. 
        Return type is UBASE.
        """
        typeName = self.actualType(tpe, system)
        if typeName in self._map.keys():
            unitName = self.fullName(system, name)
            fullTypeName  = self.fullType(tpe,system)
            return self._map[typeName].unit(fullTypeName , unitName)
        else:
            return UBASE.UBASE()
    
    def unitNames(self, *inp):
        """
        Get a list of unit names in the conversion. 
        If no arguments (or more than two) are supplied a 
        list of all unit names is returned.
        
        Parameters
        ----------
        inp[0] type   - (input, string) the unit type.
        inp[1] system - (input, string) the unit system.

        Return
        ------
        A list of unit names in the conversion.
        Return type is a list of strings.
        """
        if len(inp) == 2:
            typ = inp[0]
            system = inp[1]
            lst = list()
            actType = self.actualType(typ, system)
            if actType in self._map.keys():
                us = self._map[actType]
                names = us.unitNames(self.localType(typ))
                for n in names:                    
                    if self.checkName(system,n):
                        continue
                    else:
                        lst.append(self.localName(typ,n))
                return lst
            else:
                return lst
        else:
            lst = list()
            for item in self._map.items():
                us = item[1]
                types = us.typeNames();
                for itl in types:
                    names = us.unitNames(itl)
                    for i in names:
                        if i in lst:
                            continue
                        else:
                            lst.append(self.localName(itl, i))
            return lst
    
    def valid(self):
        """
        Get the validity of the conversion.

        Return
        -------
        Boolean: true if the conversion is valid, false otherwise.
        Return type is boolean.
        """
        return self._valid  
# EOF